django 1.3 compatibility with dynamic hashers from 1.4

diox opened the pull request:
If a project uses django_sha2 + django 1.4 + dynamic hashers generation as suggested in the README,  and then goes back to 1.3, checking the password of any user whose password was "upgraded" to the new-style hashers fails.

If I understand the code correctly, this is because the hashers used for 1.4 support have different prefixes, one for each entry in HMAC_KEYS, which wasn't the case before. In check_password() monkeypatch, there is a check to see if django_sha2 should use its own bcrypt_auth.check_password() code but it only checks if the hasher is "bcrypt" or "hh", so it fails if any dynamic hasher was used. 

This pull request makes the check_password() monkeypatch also check if the hasher is in the dynamic hashers list. I haven't included tests yet because I'm not sure this is the right solution at the moment and I'd prefer having feedback on the patch before working on some tests.

diox committed:
Add compatibility with 1.4 dynamic hashers even in 1.3, and fix HMAC_KEYS order in 1.4

rik commented:
Hey @fwenzel can you take a look at this?

fwenzel commented on django_sha2/auth.py:
```diff
@@ -63,7 +65,7 @@ def check_password(self, raw_password):
         """
         hashed_with = self.password.split('$', 1)[0]
-        if hashed_with in ['bcrypt', 'hh']:
+        if hashed_with in ['bcrypt', 'hh'] or hashed_with in get_dynamic_hasher_names(settings.HMAC_KEYS):
```
Line-break after the "or".

fwenzel commented:
Oh wow, how did I miss this pull request? Thanks for the PR @diox, good work! (And thanks @Rik for pointing me back here).

I think the code is good! @diox will you write some tests and add them to the PR?

Again, sorry for the silly delay :-/

diox commented:
Sure, working on that now.

diox commented:
@fwenzel : Corrected the pep8 issues and added a test.

While doing the tests, I stumbled on a couple weird issues : 
- HMAC_KEYS was not sorted so wasn't used correctly in 1.4 sometimes, depending on how the dict was set up and how python was returning it. I modified the settings to demonstrate the issues, and fixed it as well.
- Also, Sha2Tests were failing in django 1.3 (see issue #15) and I did a quick fix while I was at it. Sorry if that pollutes the PR, this was annoying me when running the tests.

I think everything is all right now, but a second pair of eyes would be nice.

fwenzel commented on test/django13/tests/test_bcrypt.py:
```diff
@@ -42,6 +42,37 @@ def test_nokey(self):
+        }
+
+        u = User.objects.get(username="john")
+        django14_style_password = prefix + raw_hashes['john'] + suffix
```
Chaining of "+" to concatenate text is frowned upon in Python. You want `''.join((my, tuple))`.

fwenzel replied:
or `'%s%s%s' % (my, other, tuple)` of course ;)

fwenzel commented:
Good job, this looks good! Will you squash your commits into a single commit? (`git rebase -i HEAD~7` or so and then squash them together).

diox commented:
Done ! The string concatenation is gone, all tests are passing and I rebased everything in one commit.

fwenzel commented:
You, sir, are a scholar and a gentleman. Thank you!
